File: Java/src/NonLazy/BankersQueue.java
package NonLazy;

import Interfaces.List;
import Interfaces.Queue;

import java.util.Iterator;

public class BankersQueue<T> implements Queue<T>
{
    private int sizef;

    private List<T> front;

    private int sizer;

    private List<T> rear;

    private BankersQueue(int sizef, List<T> front, int sizer, List<T> rear)
    {
        this.sizef = sizef;
        this.front = front;
        this.sizer = sizer;
        this.rear = rear;
    }

    public BankersQueue()
    {
        front = new PersistentList<>();
        rear = new PersistentList<>();
    }

    private BankersQueue<T> check(int sizef, List<T> front, int sizer, List<T> rear) {
        if (sizer <= sizef) return new BankersQueue<>(sizef, front, sizer, rear);
        return new BankersQueue<>(sizef + sizer, front.concat(rear.reverse()), 0, new PersistentList<>());
    }

    @Override
    public boolean isEmpty()
    {
        return sizef == 0;
    }

    @Override
    public Queue<T> snoc(T value)
    {
        return check(sizef, front, (sizer+1), rear.prepend(value));
    }

    @Override
    public T head()
    {
        if (front.isEmpty()) throw new RuntimeException("empty queue");
        return front.head();
    }

    @Override
    public Queue<T> tail()
    {
        if (front.isEmpty()) throw new RuntimeException("empty queue");
        return check(sizef-1, front.tail(), sizer, rear);
    }

    @Override
    public Iterator<T> iterator() {
        return new BankersQueueIterator<>(this);
    }


    private static final class BankersQueueIterator<T> implements Iterator<T> {

        private BankersQueue<T> bankersQueue;

        private boolean firstList = true;

        private Iterator<T> listIterator;

        BankersQueueIterator(BankersQueue<T> bankersQueue) {
            this.bankersQueue = bankersQueue;
            if (bankersQueue != null && !bankersQueue.isEmpty()) {
                listIterator = bankersQueue.front.iterator();
            }
        }

        @Override
        public boolean hasNext() {
            if (bankersQueue != null && !bankersQueue.isEmpty() && listIterator != null) {
                if (listIterator.hasNext()) return true;
                if (firstList) {
                    firstList = false;
                    listIterator = bankersQueue.rear.reverse().iterator();
                    return listIterator.hasNext();
                }
            }
            return false;
        }

        @Override
        public T next() {
            return listIterator.next();
        }
    }
}

File: Haskell/BankersQueue.hs
module BankersQueue (BankersQueue) where
    import Prelude hiding (head, tail)
    import Queue

    data BankersQueue value = BQ Int [value] Int [value]

    -- lenf/lenr is the length of the front/rear list
    -- check makes sure that the rear is added to the front if it is larger than the front
    check lenf front lenr rear =
        if lenr <= lenf 
        then BQ lenf front lenr rear
        else BQ (lenf + lenr) (front ++ reverse rear) 0 []

    instance Queue BankersQueue where
        empty = BQ 0 [] 0 []
        isEmpty (BQ lenf _ _ _) = lenf == 0
        -- the value is prepended to the rear and the result is checked
        snoc (BQ lenf front lenr rear) value = check lenf front (lenr + 1) (value : rear)

        head (BQ _ [] _ _) = error "empty queue"
        head (BQ _ (value : front) _ _)  = value
        
        tail (BQ _ [] _ _) = error "empty queue"
        tail (BQ lenf (value : front) lenr rear) = check (lenf - 1) front lenr rear
